drivers/SmartThings/matter-thermostat/src/init.lua

```diff
@@ -283,13 +458,166 @@ local function do_configure(driver, device)
+  [capabilities.tvocMeasurement.NAME] = units.PPM
+}
+
+-- All ConcentrationMesurement clusters inherit from the same base cluster definitions,
+-- so CarbonMonoxideConcentratinMeasurement is used below but the same enum types exist
+-- in all ConcentrationMeasurement clusters
+local level_strings = {
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.UNKNOWN] = "unknown",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.LOW] = "good",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.MEDIUM] = "moderate",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.HIGH] = "unhealthy",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.CRITICAL] = "hazardous",
+}
+
+local molecular_weights = {
+  [capabilities.formaldehydeMeasurement.NAME] = 30.031,
+}
+
+local conversion_tables = {
+  [units.PPM] = {
+    [units.PPM] = function(value) return utils.round(value) end,
+    [units.UGM3] = function(value, molecular_weight) return utils.round((value * molecular_weight * 10^3) / MGM3_PPM_CONVERSION_FACTOR) end,
+    [units.MGM3] = function(value, molecular_weight) return utils.round((value * molecular_weight) / MGM3_PPM_CONVERSION_FACTOR) end,
+  },
+  [units.PPB] = {
+    [units.PPM] = function(value) return utils.round(value/(10^3)) end
+  },
+  [units.PPT] = {
+    [units.PPM] = function(value) return utils.round(value/(10^6)) end
+  },
+  [units.MGM3] = {
+    [units.UGM3] = function(value) return utils.round(value * (10^3)) end,
+    [units.PPM] = function(value, molecular_weight) return utils.round((value * MGM3_PPM_CONVERSION_FACTOR) / molecular_weight) end,
+  },
+  [units.UGM3] = {
+    [units.UGM3] = function(value) return utils.round(value) end,
+    [units.PPM] = function(value, molecular_weight) return utils.round((value * MGM3_PPM_CONVERSION_FACTOR) / (molecular_weight * 10^3)) end,
+  },
+  [units.NGM3] = {
+    [units.UGM3] = function(value) return utils.round(value/(10^3)) end
+  },
+  [units.BQM3] = {
+    [units.PCIL] = function(value) return utils.round(value/37) end
+  },
+}
+
+local function unit_conversion(value, from_unit, to_unit, capability_name)
+  local conversion_function = conversion_tables[from_unit][to_unit]
+  if conversion_function == nil then
+    log.info_with( {hub_logs = true} , string.format("Unsupported unit conversion from %s to %s", unit_strings[from_unit], unit_strings[to_unit]))
+    return 1
+  end
+
+  if value == nil then
+    log.info_with( {hub_logs = true} , "unit conversion value is nil")
+    return 1
+  end
+
+  -- checks if number of parameters is greater than 1. This supports conversions
+  -- where there is a value other than simple scaling values
+  local conversion_function_info = debug.getinfo(conversion_function, "u")
+  if conversion_function_info.nparams > 1 then
+    return conversion_function(value, molecular_weights[capability_name])
```

Comment:
You can just call `conversion_function(value, molecular_weights[capability_name])` and any extra params are ignored in lua. However, if this were called for a concentration measurement that isn't `formaldehydeMeasurement` this would result in a driver crash due to a nil value being used for the molecular weight.

You could add the molecular weights for every concentration to the `molecular_weights` table to address this.

Reply:
Would that be necessary, or can we just leave this as is?

Reply:
done